src/main/scala/lms/collection/PointerOps.scala

```diff
@@ -0,0 +1,42 @@
+package lms.collection
+
+import lms.core._
+import lms.util._
+import lms.core.stub._
+import lms.core.Backend._
+import lms.core.virtualize
+import lms.core.utils.time
+import lms.macros.SourceContext
+import scala.collection.mutable.HashMap
+
+trait PointerOps { b: Base =>
+  object Pointer {
+    def apply[A: Manifest](x: Var[A])(implicit pos: SourceContext) = {
+      val a = Wrap[Pointer[A]](Adapter.g.reflect("pointer-new", UnwrapV(x)))
+      cache += ((Unwrap(a), UnwrapV(x)))
+      a
+    }
+    def deref[A: Manifest](x: Rep[Pointer[A]])(implicit pos: SourceContext) = {
```

Comment:
I think `deref` is better to be a method of something like `PointerOps` (which provides the syntax and operations you can do on a pointer instance), instead of in the singleton object `Pointer`.

Reply:
make sense. However, it is actually designed to not be used by users (only used to get the accurate read/write effects to the correct Sym). Does this make sense? Maybe it should be made private somehow?